### PR TITLE
feature: allow concurrent runs of the same workflow instance

### PR DESCRIPTION
**Pull request by masci**
# Description

Move the data structures used to implement the workflow runtime into a new object `WorkflowSession`, decoupling it from `Workflow` and allowing mulitple instance of the runtime (read: concurrent `run()`s).

Example:
```py
class DummyWorkflow(Workflow):
    @step()
    async def step(self, ev: StartEvent) -> StopEvent:
        return StopEvent(result=ev.number * 2)

workflow = DummyWorkflow()
results = await asyncio.gather(
    workflow.run(number=3), workflow.run(number=42), workflow.run(number=-99)
)
```

Fixes https://github.com/run-llama/llama_index/issues/15455

Note this moves `self.send_event()` to `ctx.session.send_event()` and I couldn't find a way to keep the old one, so the PR is to be considered breaking.

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama-index-core/llama_index/core/workflow/session.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        self._tasks: Set[asyncio.Task] = set()
+        self._broker_log: List[Event] = []
+        self._step_flags: Dict[str, asyncio.Event] = {}
+        self._accepted_events: List[Tuple[str, str]] = []
+        self._retval: Any = None
+        self._root_context = Context(self)
+        # Context management
+        self._step_to_context: Dict[str, Context] = {}
+
+    def send_event(self, message: Event, step: Optional[str] = None) -> None:
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/workflow/session.py:28]: so if step is not None, we are sending the event only to that step (assuming it accepts that event?)

Comment (masci) [reply llama-index-core/llama_index/core/workflow/session.py:28]: Yes, this is from https://github.com/run-llama/llama_index/pull/15259

**Comment by logan-markewich** [comment]
@masci hmm, with this change, does that mean that the context is no longer persistent between runs? Or each run shares the same context? Having trouble tracing the code path there

**Comment by masci** [comment]
> @masci hmm, with this change, does that mean that the context is no longer persistent between runs? Or each run shares the same context? Having trouble tracing the code path there

Every session has a different context. Before, the root context was attached to the workflow instance, not is attached to the workflow session. For now I keep all the sessions that have been spawned in `self._sessions` but are not used, the door is open for "resuming" a session (but that would make the UX way more complex).

**Comment by logan-markewich** [comment]
@masci so the consequence here, is that each `.run()` call could be considered stateless? Whereas before every `.run()` call was stateful because the context was attached/modified at the root?

**Comment by jerryjliu** [comment]
> @masci so the consequence here, is that each `.run()` call could be considered stateless currently (barring a way to resume sessions)? Whereas before every `.run()` call was stateful because the context was attached/modified at the root?

without reading the code deeply, i think this makes sense to me? especially if a user is serving multiple user requests
- you could still store attributes in `__init__`
- i'm not sure how easy it is to pass in + get state from a given session. could be interesting to think about state management per user/session as its own abstraction, separate from a given workflow

**Review by logan-markewich — APPROVED** [review]
Alright, I'm on board with this. We can patch up docs as needed in a followup PR

**Pull request merged**